Update Repository Readme (#6387)
ScratchAddons/ScratchAddons: README.md

[![](https://img.shields.io/github/v/release/ScratchAddons/ScratchAddons?style=flat-square&logo=github&logoColor=white&label=GitHub&color=181717)](https://github.com/ScratchAddons/ScratchAddons/releases)

[![](https://img.shields.io/github/license/ScratchAddons/ScratchAddons?style=flat-square)](https://github.com/ScratchAddons/ScratchAddons/blob/master/LICENSE)
[![](https://img.shields.io/badge/chat-on_discord-7289da.svg?style=flat-square)](https://discord.gg/R5NBqwMjNc)
[![](https://img.shields.io/badge/website-scratchaddons.com-ff7b26.svg?style=flat-square)](https://scratchaddons.com)

## About

Scratch Addons combines new and existing features and themes for the [Scratch](https://scratch.mit.edu) website and project editor into one easy-to-access and configurable browser extension. The mission is to provide a centralized, up-to-date platform for community development of new features and themes for Scratch.

<div align="center">

![Scratch Addons screenshot](.github/readme-images/screenshot.png)

</div>

> ### What is an addon?

Reviewer: ```suggestion
### About addons
```

> An addon is one or more [userscripts](https://scratchaddons.com/docs/develop/userscripts/) or [userstyles](https://scratchaddons.com/docs/develop/userstyles/) that run on the Scratch website or project editor.

Reviewer: ```suggestion
An addon is one or more [userscripts](https://scratchaddons.com/docs/develop/userscripts/) (written in JavaScript) or [userstyles](https://scratchaddons.com/docs/develop/userstyles/) (written in CSS) that run on the Scratch website or project editor.
```

> Userscripts can use the `addon.*` JavaScript APIs provided by Scratch Addons, which allows them to obtain Scratch-related information like the currently logged in user and use extension APIs to do things such as send notifications.

Reviewer: I must admit the `addon.* APIs` description on the docs is quite old. Only one addon sends notifications and not through an addon.* API.

```suggestion
Userscripts can also use the `addon.*` JavaScript APIs provided by Scratch Addons, which make it easy to obtain Scratch-related information like the currently logged in user, wait until a certain element exists on the page, or get a reference to the Scratch VM object.
```

Reviewer: Even better if you modify my suggestion to imply the addon.* API can do many more things than the three ones I mentioned.

Read the [documentation](https://scratchaddons.com/docs/develop/getting-started/addon-basics/) for more information about addons.

### If everything is an addon, what does Scratch Addons do?

> Scratch Addons by itself is just an addon loader. Its main tasks are to:

Reviewer: I would remove this whole section until we have better docs on non-addon code, in that case we'll just link to them.

Reviewer: Maybe this section could be renamed to `About the browser extension` and it can be made shorter.

Reviewer: Maybe we could go straight to the point: the browser extension only runs addons when necessary (if the URL matches), it provides addon.* APIs and provides a settings page to the user.

## About

Scratch Addons combines new and existing features and themes for the [Scratch](https://scratch.mit.edu) website and project editor into one easy-to-access and configurable browser extension. The mission is to provide a centralized, up-to-date platform for community development of new features and themes for Scratch.

<div align="center">

![Scratch Addons screenshot](.github/readme-images/screenshot.png)

</div>

### About addons

An addon mainly consists of one or more [userscripts](https://scratchaddons.com/docs/develop/userscripts/) (written in JavaScript) or [userstyles](https://scratchaddons.com/docs/develop/userstyles/) (written in CSS) that run on the Scratch website or project editor.

Each addon declares its own [addon manifest](https://scratchaddons.com/docs/reference/addon-manifest/) (`addon.json` file). This file specifies under which circumstances each one of its userscripts and userstyles should be injected into the page. It also contains user-facing information, such as the description of the feature, and more information about the addon's settings.

> Userscripts work similarly to [extension content scripts](https://developer.chrome.com/docs/extensions/content_scripts/), but they also have access to `addon.*` APIs. Userscripts can use these built-in utilities for various purposes: waiting until a certain element exists on the page, listening to URL change events, getting a reference to the Scratch VM object, etc.

Reviewer: I may try to rewrite this but keeping the content scripts parallelism.

Other open-source projects (such as Scratch forks) can also make use of the addons. For example, the [TurboWarp](https://github.com/TurboWarp/scratch-gui/tree/develop/src/addons) and [Adacraft](https://gitlab.com/adacraft/scratch-mod/scratch-gui/-/tree/adacraft/main/src/addons) project editors allow you to use most of the editor addons without installing browser extensions, and even while offline ([TurboWarp Desktop](https://desktop.turbowarp.org/)).

### File structure

#### Addons
- `addons-l10n`: Translation for addon strings (one file per addon).
- `addon-api`: Implementation of the `addon.*` JavaScript APIs.
- `addons`: Each addon has its own directory, which must include an addon manifest file named `addon.json`.
- `libraries`: Third-party libraries and other utilities, some of which are used by addons.

#### Others
- `.github`: GitHub templates, workflows and contributing files.
- `_locales`: Translation strings for the browser extension (excludes addons).
- `background`: Background scripts for the extension.
- `content-scripts`: Content scripts, which among other things, execute userscripts and inject userstyles to the page.

> - `images`: Logos, screenshots and icons. Addon-specific images belong in the addon's directory instead.

Reviewer 2: Here as well

- `popups`: Addon pages that are only accessible through the extension popup (for example, Scratch Messaging).
- `webpages`: The settings page, extension popup, and other pages.

## Installation

> Building is not required unless working with the `webpages` directory. The best way to download the source is with Git:

Reviewer 2: We do not build webpages for now.

```sh
git clone https://github.com/ScratchAddons/ScratchAddons.git
```

For browser support information and other installation methods, check [the documentation](https://scratchaddons.com/docs/getting-started/installing).

### Loading the extension (Chrome)

To load the extension into most Chromium-based browsers, go to `chrome://extensions`, turn on developer mode, click "Load unpacked", and select the `ScratchAddons` folder.

### Loading the extension (Firefox)

> Go to `about:debugging`, select "This Firefox", click "Load Temporary Add-on..." and select the `manifest.json` file in the `ScratchAddons` folder.

Reviewer 2: oxford comma

<img src="images/icon.svg" alt="Scratch Addons logo" align="right" width="128px"></img>

> # Scratch Addons Browser Extension

Reviewer 3: For the not-sure-how-many times here, I would still suggest this.

```suggestion
# Scratch Addons
```

Reviewer 3: Just want to tell that I don't intend malice or being forceful here. If you think that's the case, I don't mean to. Sorry.

Author: I didn't mean to sound rude.

Reviewer: It's okay, if someone cares enough about changing the title they can make a dedicated PR for that after we merge this one

[![Chrome Web Store](.github/readme-images/cws-badge.png)](https://chrome.google.com/webstore/detail/fbeffbjdlemaoicjdapfpikkikjoneco)

> [![Firefox Add-ons](.github/readme-images/ff-addon-badge.png)](https://chrome.google.com/webstore/detail/fbeffbjdlemaoicjdapfpikkikjoneco)

Reviewer 4: Looks like both of these links go to the Chrome Web Store.
```suggestion
[![Firefox Add-ons](.github/readme-images/ff-addon-badge.png)](https://addons.mozilla.org/firefox/addon/scratch-messaging-extension/)
```

Each addon declares its own [addon manifest](https://scratchaddons.com/docs/reference/addon-manifest/) (`addon.json` file). This file specifies under which circumstances each one of its userscripts and userstyles should be injected into the page. It also contains user-facing information, such as the description of the feature, and more information about the addon's settings.

Userscripts work similarly to [extension content scripts](https://developer.chrome.com/docs/extensions/content_scripts/), but they also have access to `addon.*` APIs. Userscripts can use these built-in utilities for various purposes: waiting until a certain element exists on the page, listening to URL change events, getting a reference to the Scratch VM object, etc.

Addons are designed to be compatible with each other. They are also developed with performance, internationalization, accessibility, and privacy in mind.

Read the [documentation](https://scratchaddons.com/docs/develop/getting-started/addon-basics/) for more information about addons.

### About the extension

> The Scratch Addons browser extension provides a settings page where users can enable, disable and configure addon settings. The extension interprets addon manifests, stores the user's settings, and provides APIs to userscripts. A new version of the extension is sent to stores at least once a month.

Reviewer 4: Might I suggest rewording the part where a new version is released at least once a month? I would also use the word "release" and mention how each major update releases with new features.

Reviewer: Go ahead

[![](https://img.shields.io/badge/chat-on_discord-7289da.svg?style=flat-square)](https://discord.gg/R5NBqwMjNc)
[![](https://img.shields.io/badge/website-scratchaddons.com-ff7b26.svg?style=flat-square)](https://scratchaddons.com)

<img src=".github/readme-images/demo.png" alt="Scratch Addons demo" height=500px></img>

## What is Scratch Addons?

> Scratch Addons combines new and existing features and themes for the [Scratch](https://github.com/scratchfoundation) website and project editor into one easy-to-access and configurable browser extension. The mission is to provide a centralized, up-to-date platform for community development of new features and themes for Scratch.

Reviewer 5: ```suggestion
Scratch Addons combines new and existing features and themes for the [Scratch](https://scratch.mit.edu) website and project editor into one easy-to-access and configurable browser extension. The mission is to provide a centralized, up-to-date platform for community development of new features and themes for Scratch.
```

Reviewer: Maybe https://scratch.mit.edu/about if you prefer.

- Avoid addons from interfering with each other.
- Avoid duplicate work from different addons.

### Addons beyond the Scratch Addons browser extension

> Other extensions and even forks of Scratch can also provide their users with most addons from Scratch Addons, as long as a compatibility layer for `addon.*` and other parts of the addon loader are present. A notable example is the [TurboWarp](https://github.com/TurboWarp/scratch-gui/tree/develop/src/addons) editor. These also inherit translations from the Scratch Addons project.

Reviewer: We could mention adacraft (TW fork)

Author: We link to the source for TurboWarp. Should we just to the site or editor for both?

> ### Addons outside of Scratch Addons itself

Reviewer: Maybe this could be rephrased? Something like:

```suggestion
### Addons beyond the Scratch Addons browser extension
```
